daffodil-test/src/test/resources/org/apache/daffodil/section23/dfdl_expressions/expressions3.tdml

```diff
@@ -0,0 +1,400 @@
+ xmlns:dfdl="http://www.ogf.org/dfdl/dfdl-1.0/"
+ xmlns:dfdlx="http://www.ogf.org/dfdl/dfdl-1.0/extensions"
+ xmlns:ex="http://example.com"
+ xmlns:fn="http://www.w3.org/2005/xpath-functions"
+ defaultRoundTrip="true">
+
+  <tdml:defineSchema name="s1" elementFormDefault="unqualified">
+    <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd" />
+    <dfdl:format ref="ex:GeneralFormat" lengthKind="delimited"/>
+
+    <xs:group name="g">
+      <!-- 
+        The path expression below is going to have different type
+        elements for the different instances of this group in the 
+        final schema. 
+        
+        Ultimately they all are converted to string.
+       -->
+      <xs:sequence>
+      <xs:choice dfdl:choiceDispatchKey="{ ../foo/bar }">
+        <xs:element name="c1" type="xs:string" dfdl:choiceBranchKey="1" />
+        <xs:element name="c2" type="xs:string" dfdl:choiceBranchKey="2" />
+      </xs:choice>
+      </xs:sequence>
+    </xs:group>
+
+    <xs:element name="r">
+      <xs:complexType>
+        <xs:sequence dfdl:separator="|">
+          <xs:element name="e1">
+            <xs:complexType>
+              <xs:sequence>
+                <xs:element name="foo">
+                  <xs:complexType>
+                    <xs:sequence dfdl:separator="|">
+                      <xs:element name="bar" type="xs:string" />
+                      <xs:group ref="ex:g" />
+                    </xs:sequence>
+                  </xs:complexType>
+                </xs:element>
+              </xs:sequence>
+            </xs:complexType>
+          </xs:element>
+          <xs:element name="e2">
+            <xs:complexType>
+              <xs:sequence>
+                <xs:element name="foo">
+                  <xs:complexType>
+                    <xs:sequence dfdl:separator="|">
+                      <xs:element name="bar" type="xs:int" />
+                      <xs:group ref="ex:g" />
+                    </xs:sequence>
+                  </xs:complexType>
+                </xs:element>
+              </xs:sequence>
+            </xs:complexType>
+          </xs:element>
+        </xs:sequence>
+      </xs:complexType>
+    </xs:element>
+ </tdml:defineSchema>
+
+  <tdml:parserTestCase name="test_polymorphic_expr_1" model="s1" root="r"
+    description="Polymorphic path expression">
+    <tdml:document>1|a|2|b</tdml:document>
+    <tdml:infoset>
+      <tdml:dfdlInfoset>
+        <ex:r>
+          <e1>
+            <foo>
+              <bar>1</bar>
+              <c1>a</c1>
+            </foo>
+          </e1>
+          <e2>
+            <foo>
+              <bar>2</bar>
+              <c2>b</c2>
+            </foo>
+          </e2>
+        </ex:r>
+      </tdml:dfdlInfoset>
+    </tdml:infoset>
+    <tdml:warnings>
+      <tdml:warning>should be manually cast</tdml:warning>  
+    </tdml:warnings>
+  </tdml:parserTestCase>
+
+  <tdml:defineSchema name="s2a" elementFormDefault="unqualified">
```

Comment:
Please review these test cases carefully.

In this first one, you see that to get the choiceDispatchKey expression to work, you have to do lots of manual conversions. That's because ../foo/bar path expression has different types for the two different reuses of this group.

Reply:
Can you add a test case that combines xs:float, xs:decimal, xs:double, as well as one that combines xs:anyURI and xs:string (if we even support xs:anyURI, which I am not sure we do).

According to XPath 2.0, float and decimal are both autocast to double, and anyURI can be autocast to string.

I am not clear on what the implications of this for us are. In the expression "{ ../foo/bar + 1 }", it is not clear to me what type the addition should take place in. I believe our current behaviour is that the literal "1" is cast to whichever type bar has in the given context, which is then the resulting type of the entire expression (although I have been wrong about our coercien rules before). To make this work polymorphically, we would need to decide what type we want ../foo/bar to be, and insert the auto-casting, which would slightly change the sementics at the call-sights which end up relying on the cast. In fact, we might have a simmiliar problem when ../foo/bar is a combination of different decimal types: the addition of a new context in which the expression is used may change the precision with which it is computed at all contexts.

Reply:
DFDL does not have anyURI type.

I do have example of a path that is shared and has type int/float, and float/complex-type. 

This is a point where the DFDL spec is unclear. It takes no position on whether expressions are shared or copied/inlined per context. What I've done is make it so that if properties are expressed on a reference term (element ref, sequence ref, or choice ref) then the properties are copied. This is a natural consequence of the fact that the properties might appear on the ref, not on the shared object being referenced, so they may in fact be distinct property values despite referencing a shared definition.

However, if an expression appears on a local element, sequence or choice, then it is inside an element decl, or group definition, so then it is not being shared, and so no copying is being done.  BUT, a relative path within it it that is upward-and-outward (e.g., ../../foo/bar) may reference outside the global decl/definition. This is the case in my test examples. In this case the path is (now, with this change set) being interpreted as a polymorphic expression ... one that must make sense in all contexts, and must have a common definition shared definition in all contexts.

So if we have int and float interpretations we will get double behavior since the LUB of int and float is SignedNumeric, and the most general representation of that (the type all such can be converted into) is double. Same with float and decimal.  But if it is int and decimal, you'll get decimal behavior since decimal is a supertype of int. 

I can add examples that show the int/decimal case, so as to show you don't always end up with double. The only way to test this in TDML is to have a test that requires an auto-coercion be added, and then detect the warning for this, or possibly set a flag that makes the lack of this auto-coercion to cause an error. The error can be scrutinized for "decimal". I will construct one of these examples.

Reply:
Tests added

Reply:
Thinking about this some more, it seems like we want to behave as-if we are inlining everything. The alternative is to have spooky action at a distance.

If we write schema with ```<xs:group ref="foo"/>``` in our schema today, then 5 years from now, extend our schema with a choice to support a new version of the format in question and re-use the group foo in a different context, we do not want that to change the behavior on any of the pre-existing messages. Similarly, if we see 2 identical group definitions in the same file, it should not change the behavior of the schema to combine them into a single group definition.

Reply:
Yes. I guess I agree. Full referential transparency should be the goal.
Whether you copy the reusable thing inline or keep it shared should not change the behavior. 

There is another alternative - parametric - require people to bind variables rather than allow these upward-and-outward paths. The parameters have types. Another equivalent possibility: require encapsulating any upward-and-outward path expression in a constructor call.. So instead of ../../foo/bar you must write xs:double( ../../foo/bar ) (or other constructor) which is another way of forcing a type-declaration - in this case the type of foo/bar must be suitable to be an arg to the double constructor, which means I believe that it just has to have a value, since those constructors even take strings. This is, I guess, just weakening the type checking substantially. 

Ultimately, keeping track of every point of usage, the types at those points of use, and generating N specializations for the N actual situations that arise, is not impossible. It's just harder. 

E.g., consider an expression with multiple paths in it, each of which is upward-and-out to a variety of different types. The number of instance you need is potentially the product of those.... 

I would probably approximate this as follows: If there is exactly one type for all usages share. Otherwise copy.  I would certainly hope the vast bulk of expressions are used monomorphically. 

Are there any other ideas for how to achieve full referential transparency here?

Reply:
Full inlining is probably sufficient. As you said, there is not much use for polymorphic expressions. The time I could see this arising would be with massive super schemas that try to parse everything (eg. a dispatch on a files magic number) and developed a modular library of components. Even in such a schema it seems unlikely, but we have no experience with such schema yet. The issue here is that a small change could cause a significant performance degradation for hard to debug reasons.

What we might want to do is emit a warning when this situation arises. If this turns out to be something people actually want to do we can deal with it then. Otherwise, these polymorphic expressions are probably mistakes.

Reply:
Ah. Good point. We can just issue a diagnostic if there is polymorphism required, and see what happens. It is certainly true that the tests I have for polymorphic expressions are 100% contrived situations. I tend to agree that allowing polymorphism is more likely to just be allowing a mistake than a true intended usage.  

This is far easier to implement so I will go with this for now.

Reply:
I have implemented Not Yet Implemented SDE for polymorphism. 

The diagnostics show the examples where the polymorphism is happening, and the line numbers and types involved.